dd/dd.3
.\"/*+
.\"***********************************************************************
.\"***  C A N A D I A N   A S T R O N O M Y   D A T A   C E N T R E  *****
.\"
.\" (c) <1998>				(c) <1998>
.\" National Research Council		Conseil national de recherches
.\" Ottawa, Canada, K1A 0R6 		Ottawa, Canada, K1A 0R6
.\" All rights reserved			Tous droits reserves
.\" 					
.\" NRC disclaims any warranties,	Le CNRC denie toute garantie
.\" expressed, implied, or statu-	enoncee, implicite ou legale,
.\" tory, of any kind with respect	de quelque nature que se soit,
.\" to the software, including		concernant le logiciel, y com-
.\" without limitation any war-		pris sans restriction toute
.\" ranty of merchantability or		garantie de valeur marchande
.\" fitness for a particular pur-	ou de pertinence pour un usage
.\" pose.  NRC shall not be liable	particulier.  Le CNRC ne
.\" in any event for any damages,	pourra en aucun cas etre tenu
.\" whether direct or indirect,		responsable de tout dommage,
.\" special or general, consequen-	direct ou indirect, particul-
.\" tial or incidental, arising		ier ou general, accessoire ou
.\" from the use of the software.	fortuit, resultant de l'utili-
.\" 					sation du logiciel.
.\"
.\"***********************************************************************
.\"
.\" FILENAME
.\" dd.1
.\"
.\" PURPOSE:
.\" Describe the data dictionary.
.\"
.\"
.\"INDENT-OFF*
.\" $Log: not supported by cvs2svn $
.\" Revision 1.1.1.1  2002/02/21 20:23:33  tpaz
.\"
.\"
.\" Revision 1.1.1.1  1999/08/23 19:58:03  cvs-tuc
.\" Initial install into CVS of dhs-0.16
.\"
.\" Revision 1.3  1998/08/10 17:46:00  dunn
.\" *** empty log message ***
.\"
.\" Revision 1.2  1997/10/29 22:12:50  jaeger
.\" added macros to enhance formatting and made minor corrections.
.\"
.\" Revision 1.1  1997/09/29 14:57:56  dunn
.\" Initial revision
.\"
.\"INDENT-ON*
.\"
.\"***  C A N A D I A N   A S T R O N O M Y   D A T A   C E N T R E  *****
.\"***********************************************************************
.\"-*/
.\"
.                       \" Ty macro to start an indent for each type def.
.de Ty
.ie n .TP 14n
.el .TP 16n
..
./"
.                       \" Ss macro to start a structure definition.
.de Ss
.PP
.RS
.nf
.ft 3
.ie n .ta 12n 26n 55n
.el .ta 16n 40n 70n
..
./"
.                       \" Es macro to end a structure definition.
.\"
.de Es
.ft 1
.fi
.RE
.PP
..
.\"
.TH dd 1 29/09/97 "CADC"


.SH NAME
dd, ddAttribGet, ddExit, ddGetNext, ddInit, ddKeywordGet
\- Data dictionary access routines.


.SH SYNOPSIS
.LP
.nf
#include "gen_types.h"		
#include "gen_eptr.h"
#include "gen_str.h"
#include "gen_msg.h"

#include "dd.h
.ft
.fi
.LP
.\".nf
.\".ta 0.8i
.TP 7n
void
ddAttribGet( char *
.IR instrument,
char *
.IR keyword,
DD_ROW **
.IR row,
DD_STATUS
.IR *ddStatus
);
.TP 7n
void
ddExit( DD_STATUS
.IR *ddStatus
);
.TP 7n
void
ddGetNext( int
.IR counter,
DD_STATUS
.IR *ddStatus
);
.TP 7n
void
ddInit( DD_STATUS
.IR *ddStatus
);
.TP 7n
void
ddKeywordGet( char *
.IR instrument,
char *
.IR attribute,
DD_ROW **
.IR row,
DD_STATUS
.IR *ddStatus
);


.SH TYPEDEFS
The following datatypes are defined in the dd.h header file.
.Ty
DD_CMTS_LEN
Maximum length of comments, 73 characters.

.Ty
DD_INSTR_LEN
Maximum length of the instrument keyword, 71 characters.

.Ty
DD_KEYWORD_LEN
Maximum length of a keyword, 9 characters.

.Ty
DD_VALUE_LEN
Maximum length of a value, 71 characters.

.Ty
DD_ROW
This structure contains all information about a row from the data dictionary.
oThe members of the structure are:
.Ss
char      ddrInstr[DD_INSTR_LEN];    /\(** Instrument name.      \(**/
char      *ddrAttrib;                /\(** Attribute name.       \(**/
char      ddrKeyword[DD_KEYWORD_LEN];/\(** Keyword name.         \(**/
DHS_DATA_TYPE
          ddrType;               /\(** Data type.            \(**/
int       ddrGemMand;                /\(** Gemini mandatory.     \(**/
char      ddrNullVal[DD_VALUE_LEN];  /\(** Null value.           \(**/
char      ddrArray;                  /\(** Array type.           \(**/
char      *ddrColName;               /\(** Column name.          \(**/
char      ddrCmts[DD_CMTS_LEN];      /\(** Comments.             \(**/
.Es

.SH DESCRIPTION
.B dd
library provides management of the data dictionary.  The library has its own
configuration file which contains the data dictionary.

Following is a list of the library's functions:
.Ty
ddAttribGet()
This function searches for a row in its data dictionary given the 
.B instrument
and a FITS 
.B keyword 
name.  If a match is not found then an instrument wildcard of "*"
is used in the search.  If found, the row is returned
in the DD_ROW structure passed in.
.Ty
ddExit()
This function is used to cleanup and exit the data dictionary.
.Ty
ddGetNext()
This function returns a row from its data dictionary given the passed in
counter.  This allows you to cycle through each row in the data dictionary.
The list is sorted by the keyword and starts at 0.
The row from the data dictionary is returned in the DD_ROW structure
passed in, or NULL if the end of the list is reached.
.Ty
ddInit()
This function is used to initialize and read the data dictionary's 
configuration file.
.Ty
ddKeywordGet()
This function searches for a row in the data dictionary given the 
.B instrument
and an 
.B attribute 
name.  
If a match is not found then an instrument wildcard of "*"
is used in the search.  If found, the row is returned
in the DD_ROW structure passed in.


.SH FILES

.SS

The data dictionary library requires a configuration file called
.I libdd.config
which contains the data dictionary.

The data dictionary has ten columns.  The first one always has
the value "OK".  The remaining columns, starting at column two and
ending with column ten,  are as follows:

.LP
.PD 0
.TP 15
.I Instrument
The instrument used to acquire the data.  If the instrument is not
known then "*" is used.
.TP
.I Attribute
This is the attribute name used in the 
.B SDS 
structure.
.TP
.I Type
This is the data type and can be one of;
.B CHAR, STRING, BOOLEAN, BYTE,	UBYTE, SHORT, USHORT, INT, UINT,
.B FLOAT, DOUBLE,
or
.B ANY.  ANY
indicates that the data type could be of any type.
.TP
.I Keyword
This is the keyword that is found in or will be written in the 
.B FITS
file.
.TP
.I Gemini Mandatory
Has a value of either
.B T,
keyword is required in the primary header unit of a 
.B FITS
file, or
.B F,
keyword is not required in the primary header unit of a 
.B FITS
file.  This value is used by 
.B dhsData
to determine if this keyword should be found in the main header unit of a given
.B FITS
file.  For now it is best to use a value of F.
.TP
.I Null Value
This is used to check if the value in the 
.B FITS 
file is null.
.TP
.I Array Values
Indicates if the variable is multi-dimensional.  Valid values are
.B EXTEN,
a multi-dimensional array,
.B INDEX, an indexed array, or
.B NONE,
not an array.  Multi-dimensional arrays will be written to
.B FITS 
files as extension and indexed arrays become sequences of
keywords.
.TP
.I Column Name
Either 
.B NULL
or a 
.B column name.
This is used by 
.B dhsData 
to indicate the actual name of the column heading
in the database where, if the keyword is present in the 
.B FITS
file and not null,the data will be stored.  This column must exist in the
table used by the fh library (refer to the fh man page).
.TP
.I Comments
The comments to be written into the 
.B FITS 
file.

.PP
There are also a number of rules that must be adhered to in
the data dictionary.  They are as follows:

.IP 1. 
If the
.I Type 
is set to
.B ANY, 
then the 
.I Column Name 
must be set to "NULL"
.IP 2. 
If the
.If the
.I Type
is
.B ANY, 
then the
.I Null Value
must be set to "NULL"
.IP 3. 
If
.I Array Vaules
is not
.B NONE,
then the
.I Column Name
must be "NULL"
.IP 4.
The 
.I Instrument/Keyword
and
.I Instrument/Attribute 
pairs must be unique.
No two lines in the data dictionary can have the same values for the both
the 
.I Instrument 
and 
.I Keyword 
nor both the 
.I Instrument
and 
.I Attribute.
.IP 5.
.I Keyword 
values must be in upper-case to comply with 
.B FITS 
standards.
.IP 6.
If two rows have the same
.I Column Name
value then they must also have the same
.I Type.
.IP 7.
The table used by the 
.B dhsData
must exist with the columns listed in the
.I Column Name
values prior to starting

The following is an example of a portion of a configuration file.
(Text followed by a '#' is treated as a comment and ignored by
.B dd.

.sp
.nf

#
#  Data Dictionary
#  Instrument: An instrument name found in FITS, or the wildcard *
#  Attribute : The attribute name in an SDS structure.
#  Data Type: Valid values are:CHAR, STRING, BOOLEAN, BYTE, UBYTE, SHORT,
#     USHORT, INT, UINT, FLOAT, DOUBLE, or ANY which indicates not to check 
#     the data type as it can change.
#  Keyword: The keyword in the FITS file.
#  Gemini Mand.:  T or F.  This is for future use, set to F for now.
#  Null Value:  This is the value in the FITS or SDS struct that indicates it
#     is a null value.
#  Array: Valid values are: EXTEN, INDEX, NONE.  Which indicates if the
#     data is in a multi-dimensional array(EXTEN), an indexed array(INDEX),
#     or a scalar value(NONE).  Multi-dimensional arrays will be written in the
#     FITS file as FITS extensions, indexed arrays as a series of keywords, and
#     scalar values as a single keyword.
#  Column Name: Valid values are "NULL" (which indicates no column name), or a 
#     string name.  If it has a name, then when a FITS header is ingested, 
#     this keyword's value will be written to the database.
#  Comment: A quote enclosed comment.
#
#
#  Instrument	Attribute	Data	Keyword		 Gemini Mand.	Null	Array	Column	Comment	
#		Name in SDS	Type	Name in Fits	Keyword in Fits	Value		Name
#  ----------	---------	------	------------	-------------	-----	-----	-------	---------------------------------
#
OK FOC 		BSCALE   	DOUBLE 	BSCALE   	 	F	1.0	NONE	NULL	"REAL = TAPE*BSCALE + BZERO"
OK FOC 		BZERO    	DOUBLE 	BZERO    	 	F	0.0	NONE	NULL	""
OK FOC 		OBJECT   	STRING 	OBJECT   	 	F	NULL	NONE	NULL	"Object ID"
OK FOC 		ORIGIN   	STRING	ORIGIN   	 	F	NULL	NONE	NULL	""
OK FOC 		DATE     	STRING  DATE     	 	F	NULL	NONE	NULL	""
OK FOC 		IRAFNAME 	STRING	IRAFNAME 	 	F	NULL	NONE	NULL	"NAME OF IRAF IMAGE FILE"
OK FOC 		IRAF-MAX 	FLOAT  	IRAF-MAX 	 	F	NULL	NONE	NULL	"DATATA MAX"
OK FOC 		IRAF-MIN 	FLOAT  	IRAF-MIN 	 	F	NULL	NONE	NULL	"DATA MIN"
OK FOC 		IRAF-BPX 	INT    	IRAF-BPX 	 	F	16	NONE	NULL	"DATA BITS/PIXEL"
OK FOC 		IRAFTYPE 	STRING  IRAFTYPE 	 	F	SHORT	NONE	NULL	"PIXEL TYPE"
OK FOC 		CRVAL1   	DOUBLE 	CRVAL1   	 	F	NULL	NONE	NULL	"RA at Ref pix in decimal degrees"
OK FOC 		CRVAL2   	DOUBLE 	CRVAL2   	 	F	NULL	NONE	NULL 	"DEC at Ref pix in decimal degrees"
OK FOC 		CRPIX1   	FLOAT  	CRPIX1   	 	F	NULL	NONE	NULL	"Refpix of first axis"
OK FOC 		CRPIX2   	FLOAT  	CRPIX2   	 	F	NULL	NONE	NULL	"Refpix of second axis"
OK FOC	 	CD1_1    	FLOAT  	CD1_1    	 	F	NULL	NONE	NULL	"CD matrix"
OK FOC 		CD1_2    	FLOAT  	CD1_2    	 	F	NULL	NONE	NULL	"CD matrix"
OK *		AXISSZ1      	INT	AXISSZ1    		F	3	NONE	NULL	"length of x axis for entire image"
OK *		AXISSZ2      	INT	AXISSZ2    		F	3	NONE	NULL	"lenght of y axis for entire image"
OK *		CHUNK		BOOLEAN	CHUNK   		F	3	NONE	NULL	"indicates if Fits file is a 'chunk' of data"
OK *		multiArray	ANY	ARRAY			F	NULL	EXTEN	NULL	"Multi-dimension array for testing."
OK *		EXTLEVEL	INT	EXTLEVEL		F	NULL	NONE	NULL	"level of the extension file in a hierarchy"
OK PSPC		BLOCKED  	BOOLEAN	BLOCKED	 		F	NULL	NONE	NULL	"FITS file may be blocked"
OK PSPC		BUNIT    	STRING	BUNIT   		F	NULL	NONE	NULL	"Units of data values"
OK PSPC		CDELT1 	 	DOUBLE	CDELT1	 		F	NULL	NONE	NULL	"Coordinate increment per pixel"
OK PSPC		CDELT2 	 	DOUBLE	CDELT2	 		F	NULL	NONE	NULL	"Coordinate increment per pixel"
OK NICMOS 	NBFCPYRT	INT	NBFCPYRT		F	NULL	NONE	NULL	"Number of brief copyright statement lines"
OK NICMOS	BFCPYRT1	STRING	BFCPYRT1		F	NULL	NONE	NULL	""
OK NICMOS	BFCPYRT2	STRING	BFCPYRT2		F	NULL	NONE	NULL	""
OK NICMOS	NFLCPYRT	INT	NFLCPYRT		F	NULL	NONE	NULL	"Number of full copyright statement lines"
OK NICMOS	FLCPYRT1	STRING	FLCPYRT1		F	NULL	NONE	NULL	""
OK NICMOS	FLCPYRT2	STRING	FLCPYRT2		F	NULL	NONE	NULL	""	
OK NICMOS	FLCPYRT3	STRING	FLCPYRT3		F	NULL	NONE	NULL	""
OK NICMOS	FLCPYRT4	STRING	FLCPYRT4		F	NULL	NONE	NULL	""	
OK NICMOS	FLCPYRT5	STRING	FLCPYRT5		F	NULL	NONE	NULL	""	

.fi

.SH ENVIRONMENT VARIABLES

The environment variable
.B LIBDD_CONFIG_DIR may be set to the path of the 
directory containing the 
configuration file. If
.B LIBDD_CONFIG_DIR
is not set, 
.B DEFAULT_CONFIG_DIR 
should be set to where the configuration file can be found.  If
that is not set then the current directory will be used.


.SH DIAGNOSTICS

none.


.SH SEE ALSO

.BR dhsData (1).
.BR sf (3).
.BR fh (3).


.SH BUGS
none.

.SH AUTHORS
Jennifer Dunn
.br
Shannon Jaeger

.br



National Research Council
.br
Herzberg Institute of Astrophysics
.br
Canadian Astronomy Data Centre




CHANGES
==============================================
   DHS-lite sybase-less cfitio3.0.6 release
==============================================

gen++/genListOf.H                disable copy constructor
dhsStatus/staChannel.C           fix definition
dhsStatus/staChannel.H           fix missing public statement
dhsData/msg.C                    fix typo
dhsGen/dhsGenTmp.C               fix read_dir
dhsData/putFits.C                implement sybase-less option
dhsCommand/cmdCommandInfo.C      fix iterator declaration
dhsCommand/cmdDhs.H              fix friend declaration
conf/cxx_lib.mk                  fix clean target
conf/cc_bin.mk                   fix clean target
conf/cxx_bin.mk                  fix clean target
conf/cc_lib.mk                   fix clean target
conf/tcl.mk                      fix clean target
press/press.c                    add missing header
fh/fh.C                          precompiler switch for outdated cfitsio
dhsQlServer/qlsFrame.C           precompiler switch for outdated cfitsio
dhsQlServer/qlsDataset.C         precompiler switch for outdated cfitsio
sf/utils.c                       fix chunk size calculation, precompiler switch for outdated cfitsio
qldp/image.C                     fix Fits_IO.h header name under skycat-3.0.1
qldp/image.H                     fix Fits_IO.h header name under skycat-3.0.1
cdIngest/ingest.c                fix weird declaration?
sf/sds2fits.c                    leave libdd.config translation bug (would break GSA etc.)
                                 catch any attempt to move to an invalid hdu if maxHdr==0
                                 enforce max keys length to avoid HIERARCH convention
                                 precompiler switch for outdated cfitsio
conf/include.mk.in               several changes to handle sybase-less dhs, outdate-cfitsio,
                                 standard location for DRAMA, EPICS, etc
configure.in                     several changes to handle sybase-less dhs, outdate-cfitsio,
                                 standard location for DRAMA, EPICS, etc
dhsData/putSds.C                 critical fix for mem mapped fits
                                 precompiler switch for sybase-less dhs
dhsStatus/staMonitor.H           precompiler switch for sybase-less
dhsStatus/staMonitor.C           precompiler switch for sybase-less
dhsData/dtsDhs.C                 precompiler switch for sybase-less
dhsData/get.H                    precompiler switch for sybase-less
dhsData/get.C                    precompiler switch for sybase-less (needs improvement)
dhsStatus/Makefile               confitional flag for sybase-less
dhsData/Makefile                 confitional flag for sybase-less
dhsData/putFits.C                precompiler switch for sybase-less
dhsData/msg.C                    removing ending .
dhsData/dbm.C                    precompiler switch for sybase-less



dhs/dd/dd.c
static char rcsid[] = "$Id: dd.c,v 1.1.1.1 2002-11-24 20:24:12 brighton Exp $";
/*
 ************************************************************************
 ****  C A N A D I A N   A S T R O N O M Y   D A T A   C E N T R E  *****
 *
 * (c) <year>				(c) <year>
 * National Research Council		Conseil national de recherches
 * Ottawa, Canada, K1A 0R6 		Ottawa, Canada, K1A 0R6
 * All rights reserved			Tous droits reserves
 * 					
 * NRC disclaims any warranties,	Le CNRC denie toute garantie
 * expressed, implied, or statu-	enoncee, implicite ou legale,
 * tory, of any kind with respect	de quelque nature que se soit,
 * to the software, including		concernant le logiciel, y com-
 * without limitation any war-		pris sans restriction toute
 * ranty of merchantability or		garantie de valeur marchande
 * fitness for a particular pur-	ou de pertinence pour un usage
 * pose.  NRC shall not be liable	particulier.  Le CNRC ne
 * in any event for any damages,	pourra en aucun cas etre tenu
 * whether direct or indirect,		responsable de tout dommage,
 * special or general, consequen-	direct ou indirect, particul-
 * tial or incidental, arising		ier ou general, accessoire ou
 * from the use of the software.	fortuit, resultant de l'utili-
 * 					sation du logiciel.
 *
 ************************************************************************
 *
 * FILENAME
 * dd/src/dd.c
 *
 * PURPOSE:
 *	The dd package instantiates the keyword/attribute dictionary 
 *	( a.k.a. data dictionary, dd ) and provides access function to 
 *	get out keyword and attribute information.
 *
 * FUNCTION NAME(S)
 *      int 	ddAttribGet	: Function to get attribute out of the dd.
 *	int	ddEptrFree	: Function to free the eptr list.
 *	int	ddExit		: Function to free the dd.
 *	int	ddInit		: Function to initialize the dd.
 *      int 	ddGetNext	: 
 *      int 	ddKeywordGet	: Function to get keyword out of dd.
 *
 *INDENT-OFF*
 * $Log: not supported by cvs2svn $
 * Revision 1.1.1.1  2002/02/21 20:23:33  tpaz
 *
 *
 * Revision 1.1.1.1  1999/08/23 19:58:03  cvs-tuc
 * Initial install into CVS of dhs-0.16
 *
 * Revision 1.15  1997/12/02 20:00:57  jaeger
 * Fixed string overruns with strcpy.
 *
 * Revision 1.14  1997/09/05 23:54:02  jaeger
 * removed remaining memory leaks.
 *
 * Revision 1.13  1997/09/05 18:45:43  jaeger
 * Fixed some memory leaks
 *
 * Revision 1.12  1997/08/29 21:59:45  dunn
 * Freed rowKey leak.
 *
 * Revision 1.11  1997/08/15 20:39:51  jaeger
 * Changed searching method to ddCompAttrib in ddKeywordGet
 *
 * Revision 1.10  1997/08/12 19:17:06  jaeger
 * Fixed bug - Used strsav instead of strcpy with ddrAttrib
 *
 * Revision 1.9  1997/07/29 20:04:07  dunn
 * Lots of changes.
 *
 * Revision 1.8  1997/07/17 23:02:37  dunn
 * Fixed format msg for get attributes.
 *
 * Revision 1.7  1997/07/17 21:51:20  dunn
 * Fixed messages and added ddGetNext
 *
 * Revision 1.6  1997/07/15 18:28:47  jaeger
 * checked in for compile with dhs
 *
 * Revision 1.5  1997/07/11 22:38:09  jaeger
 * Added "ANY" to datatype
 *
 * Revision 1.4  1997/07/09 16:35:34  jaeger
 *  Converted library to use dhs instead of bdt
 *
 *INDENT-ON*
 *
 ****  C A N A D I A N   A S T R O N O M Y   D A T A   C E N T R E  *****
 ************************************************************************
*/


/*
 ************************************************************************
 *  General unix include files.
 ************************************************************************
 */

#include <string.h>


/*
 ************************************************************************
 *  CADC library include files.
 ************************************************************************
 */
 
#include "gen_types.h"		
#include "gen_eptr.h"
#include "gen_str.h"
#include "gen_msg.h"


/*
 ************************************************************************
 *  dd library include files.
 ************************************************************************
 */
#define __EDD
#include "dd.h"
#include "local_dd.h"

void	        ddFormatMessage();
void            configInit();


/*
 ************************************************************************
 *+
 * FUNCTION NAME:
 * ddAttribGet
 *
 * INVOCATION:
 * ddAttribGet( instrument, keyword, row, status );
 *
 * PARAMETERS: (">" input, "!" modified, "<" output)
 * (>) instr (char *) Instrument name.
 * (>) kw (char *) Keyword name to look in eptr list with.
 * (<) row (DD_ROW **) Data dict. entry returned.
 * (!) status (DD_STATUS *) Function return status.
 *
 * FUNCTION VALUE:
 * None
 *
 * PURPOSE:
 * Function to get attribute out of dd.
 *
 * DESCRIPTION:
 * Function to get attribute out of dd given the keyword name and instrument.
 *
 * EXTERNAL VARIABLES:
 * None.
 *
 * PRIOR REQUIREMENTS:
 * None.
 *
 * SEE ALSO:
 * None.
 *
 * DEFICIENCIES:
 * None.
 *-
 ************************************************************************
 */

 
void             ddAttribGet
(
    char       *instr,	/* (in)  Instrument 				*/
    char       *kw,	/* (in)  Keyword 				*/
    DD_ROW     **row, 	/* (out) dd entry corresponding to instr, kw 	*/
    DD_STATUS  *status  /* (mod) Function return status.                */
)
{
    int		index;
    boolean	rc;
    DD_ROW	rowKey;
 

    statCheck( *status, VOID );
 

    /* 
     *  Fill in the row_key keyword and instrument values.
     */

    strnCpy( rowKey.ddrKeyword, kw, DD_KEYWORD_LEN, *status, VOID );
    strnCpy( rowKey.ddrInstr, instr, DD_INSTR_LEN, *status, VOID );


    /*
     *  Call eptr_search with the ddCompKeyword function as an argument.
     */

    rc = eptr_search( ddKeyword, &rowKey, &index, ddCompKeyword );
    if ( rc == FALSE )
    {
	*status = DD_S_NOT_FOUND;
	ddFormatMessage( *status, "keyword", "instr/keyword",
		rowKey.ddrInstr, rowKey.ddrKeyword );
        strnCpy( rowKey.ddrInstr, "*", DD_INSTR_LEN, *status, VOID );
	rc = eptr_search( ddKeyword, &rowKey, &index, ddCompKeyword );
	if ( rc == FALSE )
	{
	    *status = DD_E_NOT_FOUND;
	    ddFormatMessage( *status, "keyword", "instr/keyword",
		    rowKey.ddrInstr, rowKey.ddrKeyword );
	    return;
	}
    }
    

    /* 
     *  Get the row out at position of index.
     */

    *row = (DD_ROW *) eptritem( ddKeyword, index );
 

    return;
}

/*
 ************************************************************************
 *+
 * FUNCTION NAME:
 * ddEptrFree
 *
 * INVOCATION:
 * Passed as parameter to eptr_free( kkKeyword, ddEptrFree );
 *
 * PARAMETERS: (">" input, "!" modified, "<" output)
 * (>) row (DD_ROW *) Row to free.
 *
 * FUNCTION VALUE:
 * None
 *
 * PURPOSE:
 * To free malloc'ed strings.
 *
 * DESCRIPTION:
 * To free malloc'ed strings.
 *
 * EXTERNAL VARIABLES:
 * None.
 *
 * PRIOR REQUIREMENTS:
 * None.
 *
 * SEE ALSO:
 * None.
 *
 * DEFICIENCIES:
 * None.
 *-
 ************************************************************************
 */

void 	ddEptrFree
(
    DD_ROW	*row		/* (in)  The row to free.		*/
)
{
    if ( row->ddrAttrib != NULL )
    {
	gen_free( row->ddrAttrib );
    }

    if ( row->ddrColName != NULL )
    {
	gen_free( row->ddrColName );
    }

    gen_free( row );
}

/*
 ************************************************************************
 *+
 * FUNCTION NAME:
 * ddExit
 *
 * INVOCATION:
 * ddExt( status );
 *
 * PARAMETERS: (">" input, "!" modified, "<" output)
 * (!) status (DD_STATUS *) Return function status.
 *
 * FUNCTION VALUE:
 * None
 *
 * PURPOSE:
 * Function to cleanup the data dictionary.
 *
 * DESCRIPTION:
 * Function to cleanup the data dictionary.
 *
 * EXTERNAL VARIABLES:
 * None.
 *
 * PRIOR REQUIREMENTS:
 * None.
 *
 * SEE ALSO:
 * None.
 *
 * DEFICIENCIES:
 * None.
 *-
 ************************************************************************
 */

void 		ddExit
(
    DD_STATUS  *status  /* (mod) function return status.                */
)
{
    statCheck( *status, VOID );
    
    /*
     * Free the eptr lists.
     * Note: ddKeyword, ddAttrib and ddColumn point to the same
     * memory space.
     */

    eptr_free( ddAttrib, ddEptrFree );
    ddAttrib = NULL;
    eptr_free( ddKeyword, NULL );
    ddKeyword = NULL;
    eptr_free( ddColumn, NULL );
    ddColumn = NULL;


    return;
}

/*
 ************************************************************************
 *+
 * FUNCTION NAME:
 * ddInit
 *
 * INVOCATION:
 * ddInit( status );
 *
 * PARAMETERS: (">" input, "!" modified, "<" output)
 * (!) status (DD_STATUS *) Return function status.
 *
 * FUNCTION VALUE:
 * None
 *
 * PURPOSE:
 * Function to initialize the data dictionary.
 *
 * DESCRIPTION:
 * Description of the algorithm.
 *
 * EXTERNAL VARIABLES:
 * None.
 *
 * PRIOR REQUIREMENTS:
 * None.
 *
 * SEE ALSO:
 * None.
 *
 * DEFICIENCIES:
 * None.
 *-
 ************************************************************************
 */

void 		ddInit
(
    DD_STATUS  *status  /* (mod) function return status.                */
)
{

    statCheck( *status, VOID );

    /*
     * Create the eptr lists.
     */

    ddAttrib = NULL;
    ddKeyword = NULL;
    ddColumn = NULL;

    nullCheck( ddKeyword = eptr_alloc( DD_LIST_CHUNK ), *status, VOID );
    nullCheck( ddAttrib = eptr_alloc( DD_LIST_CHUNK ), *status, VOID );
    nullCheck( ddColumn = eptr_alloc( DD_LIST_CHUNK ), *status, VOID );


    /*
     * Open and read the configuration file.
     */

    configInit( status );
    if ( *status != DD_S_SUCCESS ) 
    {
	return;
    }


    /* Sort the eptr lists */

    nullCheck( ddKeyword = eptr_sort( ddKeyword, ddCompKeyword ), *status,
	       VOID );
    nullCheck( ddAttrib = eptr_sort( ddAttrib, ddCompAttrib ), *status, VOID );


    return;
}

/*
 ************************************************************************
 *+
 * FUNCTION NAME:
 * ddGetNext
 *
 * INVOCATION:
 * ddGetNext ( i )
 *
 * PARAMETERS: (">" input, "!" modified, "<" output)
 * (>) i (int) Counter for eptr list.
 * (!) status (DD_STATUS) Modified return function status.
 *
 * FUNCTION VALUE:
 * Returns DD_ROW in data dictionary or NULL if you reach the end.
 *
 * PURPOSE:
 * Get data dictionary items from the eptr list.
 *
 * DESCRIPTION:
 * Get data dictionary items from the eptr list.
 *
 * EXTERNAL VARIABLES:
 * None.
 *
 * PRIOR REQUIREMENTS:
 * None.
 *
 * SEE ALSO:
 * None.
 *
 * DEFICIENCIES:
 * None.
 *-
 ************************************************************************
 */
 
DD_ROW		*ddGetNext
(       
    int	 	i,	 	/* (in)  Index into eptr list.		*/
    DD_STATUS   *status  	/* (mod) function return status.	*/
)
{
 
    statCheck( *status, NULL );
   
    if ( i >=  eptrlen( ddKeyword ) )
    {
	/*
	 *  At the end of the list.
	 */

	return( NULL );
    }

    return( eptritem( ddKeyword, i ) );
}

/*
 ************************************************************************
 *+
 * FUNCTION NAME:
 * ddKeywordGet
 *
 * INVOCATION:
 * ddKeywordGet( instrument, attribute, row );
 *
 * PARAMETERS: (">" input, "!" modified, "<" output)
 * (>) instr (char *) Instrument name.
 * (>) attrib (char *) Attribute name to look in eptr list with.
 * (<) row (DD_ROW **) Data dict. entry returned.
 * (!) status (DD_STATUS *) Function return status.
 *
 * FUNCTION VALUE:
 * none
 *
 * PURPOSE:
 * Function to get keyword out of dd.
 *
 * DESCRIPTION:
 * Function to get keyword out of dd given the attribute name.
 *
 * EXTERNAL VARIABLES:
 * None.
 *
 * PRIOR REQUIREMENTS:
 * None.
 *
 * SEE ALSO:
 * None.
 *
 * DEFICIENCIES:
 * None.
 *-
 ************************************************************************
 */
 
void             ddKeywordGet
(       
    char 	*instr, 	/* (in)  Instrument.			*/
    char 	*attrib, 	/* (in)  Attribute to look up.		*/
    DD_ROW 	**row,		/* (out) Row returned.			*/
    DD_STATUS   *status  	/* (mod) function return status.	*/
)
{
    int		index;
    boolean	rc;
    DD_ROW	rowKey;
 
    statCheck( *status, VOID );
    
    /* 
     * Fill in the row_key values.
     */

    nullCheck( rowKey.ddrAttrib = strsav( attrib ), *status, VOID );
    strnCpy( rowKey.ddrInstr, instr, DD_INSTR_LEN, *status, VOID );


    /*
     * Call eptr_search with the ddCompAttrib function as an argument.
     */

    rc = eptr_search( ddAttrib, &rowKey, &index, ddCompAttrib );
    if ( rc == FALSE )
    {
	*status = DD_S_NOT_FOUND;
	ddFormatMessage( *status, "keyword", "instr/attribute",
		rowKey.ddrInstr, rowKey.ddrAttrib );
    	strnCpy( rowKey.ddrInstr, "*", DD_INSTR_LEN, *status, VOID );
	rc = eptr_search( ddAttrib, &rowKey, &index, ddCompAttrib );
	if ( rc == FALSE )
	{
	    ddFormatMessage( DD_E_NOT_FOUND, "attribute", "instr/attribute",
		    rowKey.ddrInstr, rowKey.ddrAttrib );
	    *status = DD_E_NOT_FOUND;
	    return;
	}
    }
    

    /* 
     * Get the row out at position of index.
     */

    *row = (DD_ROW *) eptritem( ddAttrib, index );
    free( rowKey.ddrAttrib );

    return;
}
